src/main/java/com/spotify/docker/client/DefaultDockerClient.java

```diff
@@ -2736,14 +2754,63 @@ public Void call() throws DockerException, InterruptedException, IOException {
     }
   }
 
+
   private void tailResponse(final String method, final Response response,
                             final ProgressHandler handler, final WebTarget resource)
         throws DockerException, InterruptedException {
     final ExecutorService executor = Executors.newSingleThreadExecutor();
     try {
       final ProgressStream stream = response.readEntity(ProgressStream.class);
       final Future<?> future = executor.submit(
-              new ResponseTailReader(stream, handler, method, resource));
+              new ProgressResponseTailReader(stream, handler, method, resource));
+      future.get();
+    } catch (ExecutionException e) {
+      final Throwable cause = e.getCause();
+      if (cause instanceof DockerException) {
+        throw (DockerException)cause;
+      } else {
+        throw new DockerException(cause);
+      }
+    } finally {
+      executor.shutdownNow();
+      try {
+        response.close();
+      } catch (ProcessingException e) {
+        // ignore, thrown by jnr-unixsocket when httpcomponent try to read after close
+        // the socket is closed before this exception
+      }
+    }
+  }
+
+  private static class StatsResponseTailReader implements Callable<Void> {
```

Comment:
this class looks really similar to ProgressResponseTailHeader, do you think a common class could be extracted between the two to avoid the duplication?

Reply:
I have been thinking about it and saw that there is something I don't completely understand: why is a new thread created in `tailResponse`? Is it because calls to `images/create` API cannot be cancelled/interrupted? I've just copied that method but I'm not sure it applies to `stats`, as calls to the `containers/{id}/stats` API must be cancellable/interruptible.

If my thought is correct, maybe I could just call `stream.tail` directly inside `tailResponse`, instead of creating a future. What do you think? Am I missing something?

Thanks again!